core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcConnection.java

```diff
@@ -0,0 +1,439 @@
+import org.slf4j.LoggerFactory;
+
+import java.net.ConnectException;
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Copycat transport {@link Connection} implementation that uses Alluxio gRPC.
+ */
+public class CopycatGrpcConnection implements Connection, StreamObserver<CopycatMessage> {
+  private static final Logger LOG = LoggerFactory.getLogger(CopycatGrpcConnection.class);
+
+  /** Exception listeners for this connection.  */
+  private final Listeners<Throwable> mExceptionListeners;
+  /** Close listeners for this connection.  */
+  private final Listeners<Connection> mCloseListeners;
+
+  /** Whether this connection is closed already. */
+  private boolean mClosed;
+  /** Failure on the connection. */
+  private Throwable mLastFailure;
+
+  /** Counter for assigning connection bound requests Ids. */
+  private final AtomicLong mRequestCounter;
+
+  /** Map of request handlers. */
+  private final Map<Class, CopycatGrpcConnection.HandlerHolder> mHandlers;
+  /** Map of pending requests. */
+  private final Map<Long, CopycatGrpcConnection.ContextualFuture> mResponseFutures;
+
+  /** Thread context of creator of this connection. */
+  private final ThreadContext mContext;
+
+  /** Type of connection owner. */
+  private final ConnectionOwner mConnectionOwner;
+
+  /** Target's stream observer. */
+  private StreamObserver<CopycatMessage> mTargetObserver;
+
+  /** Time in milliseconds to timeout pending requests.  */
+  private final long mRequestTimeoutMs;
+
+  /** Timeout scheduler. */
+  private final Scheduled mTimeoutScheduler;
+
+  /**
+   * Creates a connection object.
+   *
+   * Note: {@link #setTargetObserver} should be called explicitly before using the connection.
+   *
+   * @param connectionOwner owner of connection
+   * @param context copycat thread context
+   * @param requestTimeoutMs timeout in milliseconds for requests
+   */
+  public CopycatGrpcConnection(ConnectionOwner connectionOwner, ThreadContext context,
+      long requestTimeoutMs) {
+    mConnectionOwner = connectionOwner;
+    mContext = context;
+    mRequestTimeoutMs = requestTimeoutMs;
+
+    mClosed = false;
+    mRequestCounter = new AtomicLong(0);
+    // Initialize timeout scheduler.
+    mTimeoutScheduler = context.schedule(Duration.ofMillis(mRequestTimeoutMs),
+        Duration.ofMillis(mRequestTimeoutMs / 2), this::timeout);
+    // Initialize listeners.
+    mExceptionListeners = new Listeners<>();
+    mCloseListeners = new Listeners<>();
+    // Initialize handler and response maps.
+    mHandlers = new ConcurrentHashMap<>();
+    mResponseFutures = new ConcurrentSkipListMap<>();
```

Comment:
Why do we need a `ConcurrentSkipListMap` for futures?

Reply:
To favor precision for timing out pending requests by iterating keys in order.